ecommerce_extensions/payment/views/ednx_pp.py

```diff
@@ -0,0 +1,209 @@
+from oscar.apps.partner import strategy
+from oscar.apps.payment.exceptions import PaymentError, TransactionDeclined
+from oscar.core.loading import get_class, get_model
+from rest_framework import status
+
+from ecommerce_extensions.payment.processors.ednx_pp import ednx_pp
+
+try:
+    from ecommerce.extensions.checkout.mixins import EdxOrderPlacementMixin
+    from ecommerce.extensions.checkout.utils import get_receipt_page_url
+    from ecommerce.extensions.payment.exceptions import InvalidSignatureError
+except ImportError:
+    class EdxOrderPlacementMixin:
+        """This a empty class in order to avoid errors in a test scenario,
+        since the dependency is not available.
+        """
+
+logger = logging.getLogger(__name__)
+
+Applicator = get_class('offer.applicator', 'Applicator')
+Basket = get_model('basket', 'Basket')
+BillingAddress = get_model('order', 'BillingAddress')
+Country = get_model('address', 'Country')
+NoShippingRequired = get_class('shipping.methods', 'NoShippingRequired')
+OrderNumberGenerator = get_class('order.utils', 'OrderNumberGenerator')
+OrderTotalCalculator = get_class('checkout.calculators', 'OrderTotalCalculator')
+
+
+class EdnxPaymentResponseView(EdxOrderPlacementMixin, View):
+    """ Validates a response from PP and processes the associated basket/order appropriately. """
+
+    @property
+    def payment_processor(self):
+        """An instance of ednx payment processor."""
+        return ednx_pp(self.request.site)
+
+    # Disable atomicity for the view. Otherwise, we'd be unable to commit to the database
+    # until the request had concluded; Django will refuse to commit when an atomic() block
+    # is active, since that would break atomicity. Without an order present in the database
+    # at the time fulfillment is attempted, asynchronous order fulfillment tasks will fail.
+    @method_decorator(transaction.non_atomic_requests)
+    @method_decorator(csrf_exempt)
+    def dispatch(self, request, *args, **kwargs):
+        return super().dispatch(request, *args, **kwargs)
+
+    def _get_billing_address(self, ednx_response):
+
+        return BillingAddress(
+            first_name=ednx_response.get('cc_holder', ''),
+            last_name=ednx_response.get('cc_holder', ''),
+            line1=ednx_response.get('billing_address', ''),
+            line4=ednx_response.get('billing_city', ''),  # Oscar uses line4 for city
+            country=Country.objects.get(iso_3166_1_a2=ednx_response['billing_country']),
+        )
+
+    def _get_basket(self, basket_id):
+        if not basket_id:
+            return None
+
+        try:
+            basket_id = int(basket_id)
+            basket = Basket.objects.get(id=basket_id)
+            basket.strategy = strategy.Default()
+            Applicator().apply(basket, basket.owner, self.request)
+            return basket
+        except (ValueError, ObjectDoesNotExist):
+            return None
+
+    def post(self, request):
```

Comment:
there are three try-except block, in order to improve the readability you can slip them in three private methods